Overflow warnings when converting RandomForestsClassifier

Hi,

Apologies if I'm just missing something obvious but when trying to convert a RandomForestsClassifier (trained on all-integer data) I keep getting tons of these warnings:
```
In file included from tmp/myinlinetree.c:2:
tmp/myinlinetree.h:62312:23: warning: overflow in conversion from ‘int’ to ‘short int’ changes value from ‘66413’ to ‘877’ [-Woverflow]
62312 |   { 6, 1736.5, 62303, 66413 },
      |                       ^~~~~
tmp/myinlinetree.h:65535:23: warning: overflow in conversion from ‘int’ to ‘short int’ changes value from ‘66017’ to ‘481’ [-Woverflow]
65535 |   { 4, 6592.5, 65526, 66017 },
      |                       ^~~~~
tmp/myinlinetree.h:65536:23: warning: overflow in conversion from ‘int’ to ‘short int’ changes value from ‘65809’ to ‘273’ [-Woverflow]
65536 |   { 3, 4978.5, 65527, 65809 },
      |                       ^~~~~
```
Any idea how I could fix that?
I was using this line to convert it, both with and without the dtype argument - same effect.
```
cmodel = emlearn.convert(forest, method='inline', dtype='int32')
```

--- a/emlearn/eml_trees.h
+++ b/emlearn/eml_trees.h
@@ -78,7 +78,12 @@ eml_trees_predict_tree(const EmlTrees *forest, int32_t tree_root,
         const float value = features[feature];
         const float point = forest->nodes[node_idx].value;
         //printf("node %d feature %d. %d < %d\n", node_idx, feature, value, point);
-        node_idx = (value < point) ? forest->nodes[node_idx].left : forest->nodes[node_idx].right;
+        const int16_t child = (value < point) ? forest->nodes[node_idx].left : forest->nodes[node_idx].right;
+        if (child >= 0) {
+            node_idx += child;
+        } else {
+            node_idx = child;
+        }
     }
 
     const int16_t leaf = -node_idx-1;

--- a/emlearn/evaluate/trees.py
+++ b/emlearn/evaluate/trees.py
@@ -25,7 +25,7 @@ def model_size_nodes(model, a=None, b=None):
     """
     em = convert_model(model)
     
-    nodes, roots = em.forest_
+    nodes, roots, leaves = em.forest_
     return len(nodes)
 
 def model_size_bytes(model, a=None, b=None, node_size=None):

--- a/emlearn/trees.py
+++ b/emlearn/trees.py
@@ -331,16 +331,25 @@ def generate_c_nodes(flat, name, dtype='float'):
         assert value >= child_value_min, value
         assert value <= child_value_max, value
 
-    def make_node(index, node):
-        feature, value, left_index, right_index = node
-
-        # XXX: consider using relative jumps?
-        left = left_index
-        right = right_index
+    def encode_child(index, child):
+        if child >= 0:
+            # decision node, use relative jump
+            assert child >= index
+            encoded = child - index
+            # must not become negative, would be confused with a leaf
+            assert encoded >= 0
+        else:
+            # leaf node, leave as-as
+            encoded = child
     
-        assert_valid_child(left)
-        assert_valid_child(right)
+        assert_valid_child(encoded)
+        return encoded
 
+    def make_node(index, node):
+        feature, value, left_child, right_child = node
+
+        left = encode_child(index, left_child)
+        right = encode_child(index, right_child)
         value = cgen.constant(value, dtype=dtype)
 
         return "{{ {}, {}, {}, {} }}".format(feature, value, left, right)
@@ -560,11 +569,6 @@ class Wrapper:
         self.n_classes = 0
         if self.is_classifier:
             self.n_classes = estimators[0].n_classes_
-
-        n_nodes = len(self.forest_[0])
-        max_nodes = 2**15 # limited by int16_t for children in EmlTreeNode structure
-        if n_nodes > max_nodes:
-            raise ValueError(f"Model has {n_nodes} nodes. Max supported is {max_nodes} nodes.")
 
 
         if classifier == 'pymodule':

--- a/test/test_trees.py
+++ b/test/test_trees.py
@@ -94,6 +94,7 @@ def huge_trees_model():
     return X, Y, est
 
 @pytest.mark.parametrize("method", ['loadable', 'inline'])
+@pytest.mark.skipif(not bool(int(os.environ.get('EMLEARN_TESTS_SLOW', '0'))), reason='EMLEARN_TESTS_SLOW not enabled')
 def test_trees_huge(method, huge_trees_model):
     """Should work just the same as a smaller model"""
     X, Y, estimator = huge_trees_model

--- a/test/test_trees.py
+++ b/test/test_trees.py
@@ -46,7 +46,7 @@ REGRESSION_DATASETS = {
     '1out': datasets.make_regression(n_targets=1, n_samples=100, random_state=random),
 }
 
-METHODS = ['pymodule', 'loadable', 'inline']
+METHODS = ['loadable', 'inline']
 
 @pytest.mark.parametrize("data", CLASSIFICATION_DATASETS.keys())
 @pytest.mark.parametrize("model", CLASSIFICATION_MODELS.keys())

--- a/test/test_trees.py
+++ b/test/test_trees.py
@@ -81,7 +81,7 @@ def test_trees_sklearn_regressor_predict(data, model, method):
 
     numpy.testing.assert_allclose(pred_c, pred_original, rtol=1e-3, atol=2)
 
-@pytest.fixture()
+@pytest.fixture(scope='module')
 def huge_trees_model():
     store_classifier_path = os.path.join(here, 'out/test_trees_huge.model.pickle')
     X, Y = datasets.make_classification(n_classes=2, n_samples=1000, random_state=1)

--- a/test/test_trees.py
+++ b/test/test_trees.py
@@ -1,8 +1,10 @@
 
+import os
+
 import sklearn
 import numpy
 import numpy.testing
-
+import joblib
 from sklearn import datasets
 from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
 from sklearn.ensemble import ExtraTreesClassifier, ExtraTreesRegressor
@@ -11,6 +13,8 @@ from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
 import emlearn
 from emlearn.evaluate.trees import model_size_nodes
 import pytest
+
+here = os.path.dirname(__file__)
 
 random = numpy.random.randint(0, 1000)
 print('random_state={}'.format(random))
@@ -77,19 +81,27 @@ def test_trees_sklearn_regressor_predict(data, model, method):
 
     numpy.testing.assert_allclose(pred_c, pred_original, rtol=1e-3, atol=2)
 
-def test_trees_too_many_nodes():
-    """should give nice error"""
+@pytest.fixture()
+def huge_trees_model():
+    store_classifier_path = os.path.join(here, 'out/test_trees_huge.model.pickle')
     X, Y = datasets.make_classification(n_classes=2, n_samples=1000, random_state=1)
-    estimator = RandomForestClassifier(n_estimators=1000, max_depth=20, random_state=1)
-    estimator.fit(X, Y)
+    est = RandomForestClassifier(n_estimators=1000, max_depth=20, random_state=1)
+    est.fit(X, Y)
 
-    with pytest.raises(ValueError) as ex:
-        cmodel = emlearn.convert(estimator, method='loadable')
+    n_nodes = model_size_nodes(est)
+    assert n_nodes >= 90*1000
 
-    error = str(ex.value).lower()
-    assert 'nodes' in error
-    assert 'model has ' in error
-    assert 'max supported ' in error
+    return X, Y, est
+
+@pytest.mark.parametrize("method", ['loadable', 'inline'])
+def test_trees_huge(method, huge_trees_model):
+    """Should work just the same as a smaller model"""
+    X, Y, estimator = huge_trees_model
+
+    cmodel = emlearn.convert(estimator, method=method)
+    pred_original = estimator.predict(X)
+    pred_c = cmodel.predict(X)
+    numpy.testing.assert_equal(pred_c, pred_original)
 
 def test_trees_to_dot():
     X, Y = datasets.make_classification(n_classes=2, n_samples=10, random_state=1)

